rust/src/tests/coordinator.rs

```diff
@@ -74,3 +74,89 @@ async fn full_cycle_1_round_1_participant() {
         }
     }
 }
+
+#[tokio::test]
+async fn dropout_1_participant_during_training() {
+    let settings = FederatedLearningSettings {
+        rounds: 1,
+        participants_ratio: 1.0,
+        min_clients: 2,
+        heartbeat_timeout: 5,
+    };
+    let (rpc_client, service_handle, _join_handle) = start_service(settings);
+
+    // Create first client. Since min_clients is 2, the heartbeat response should be `StandBy`.
+    let id_1 = service_handle.rendez_vous_accepted().await;
+    let hb_resp = service_handle.heartbeat(id_1).await;
+    assert_eq!(hb_resp, HeartBeatResponse::StandBy);
+
+    // Create second client.
+    let id_2 = service_handle.rendez_vous_accepted().await;
+
+    // Now the the round can start because the requirement of min two clients is fulfilled.
+    let round = service_handle.heartbeat_selected(id_1).await;
+    assert_eq!(round, 0);
+    let round = service_handle.heartbeat_selected(id_2).await;
+    assert_eq!(round, 0);
+
+    // Both clients start the training.
+    rpc_client
+        .mock()
+        .expect_select()
+        .returning(|_, _| future::ready(Ok(Ok(()))));
+    rpc_client
+        .mock()
+        .expect_select()
+        .returning(|_, _| future::ready(Ok(Ok(()))));
+    let (url, _token) = service_handle.start_training_accepted(id_1).await;
+    assert_eq!(&url, AGGREGATOR_URL);
+    let (url, _token) = service_handle.start_training_accepted(id_2).await;
+    assert_eq!(&url, AGGREGATOR_URL);
+
+    // Here we reset only the heartbeat of the first client and wait until the
+    // heartbeat timeout of the second client has been reached.
+    service_handle.heartbeat(id_1).await;
+    sleep_ms(2500).await;
+    service_handle.heartbeat(id_1).await;
+    sleep_ms(2500).await;
+    service_handle.heartbeat(id_1).await;
+    // The second client should be dropped.
+
+    // pretend the first client trained and sent its weights to the
+    // aggregator. The aggregator now sends an end training requests
+    // to the coordinator RPC server that we fake with the
+    // service_handle. The service should then trigger the aggregation
```

Comment:
If only one participant uploaded their weights, aggregation shouldn't be triggered. Or am I missing something?